File and range cloning support with ficlone or ficlonerange

smithdh opened the pull request:
Original feature/issue ticket is #1795. I'll add some details here to aid in review shortly.

smithdh commented:
Hi, after quesiton from @bbockelm I reaslised there's an ordering problem for any potential XrdCl plugin that would like to impliment the cloning functionality. I'll fix when I rebase & fix any merge conflicts.

My idea is to keep the plugin initalisation at Open time. But split the XrdCl::File::SetFileTemplate function at the plugin (and FileStateHandler) level into an "ExportTemplate" and "ImportTemplate" step. i.e. ExportTemplate from source File when user calls SetFileTemplate. But delay ImportTemplate on the target File until just before Open. This was already partially what the PR does; having the ExportTemplate step ensures something reasonable can be done if something strange happens, such as the source File object being destroyed between SetFileTemplate and Open.

amadio commented on src/XrdCl/XrdClFileStateHandler.cc:
```diff
@@ -3035,8 +3097,16 @@ namespace XrdCl
     req->requestid = kXR_open;
     req->mode      = self->pOpenMode;
-    req->options   = self->pOpenFlags;
+    req->options   = (self->pOpenFlags & 0xffff);
```
I wonder why this change was necessary.

amadio commented on src/XrdCl/XrdClPlugInInterface.hh:
```diff
@@ -275,6 +289,30 @@ namespace XrdCl
+      //! OpenUsingTemplatet or Clone. The returned object should be of a
+      //! class derived from ExportedFileTemplate, the exact type of which could
+      //! be used to ensure the source plugin is the same the destination. In
+      //! the case of Clone() it sould be considered that the source file object
```
There is a typo: "sould"

smithdh replied:
fixed, pushed.

smithdh commented on src/XrdCl/XrdClFileStateHandler.hh:
```diff
@@ -128,11 +144,34 @@ namespace XrdCl
       static XRootDStatus Open( std::shared_ptr<FileStateHandler> &self,
                                 const std::string                 &url,
-                                uint16_t                           flags,
+                                OpenFlags::Flags                   flags,
```
pushed change to revert type of OpenInfo::oFlags, introduced uint16_t OpenInfo::oFlags2 for upper 16 bits.

amadio commented on src/XrdCl/XrdClFileSystem.hh:
```diff
@@ -104,8 +105,10 @@ namespace XrdCl
+      IntentDirList = kXR_4dirlist, //!< Make sure the server knows we are doing
                                     //!< locate in context of a dir list operation
+      Dup      = kXR_dup<<16,       //!< Open file duplicating content from another
+      Samefs   = kXR_samefs<<16     //!< Open file on the same filesystem as another
```
I see, thanks for clarifying.

smithdh commented:
Hi Guilherme, I've added a test as an extra commit. (Should indicate success either if the cloning gives not supported, or it is successful). I also noticed a not-apparent problem with one of the new commits that we're rebased on; missing a name for kxr_clone request in the XProtocol::reqName function. So I added another commit with a fix for that (tried to also add some protection there, for missed names in the future).

I made a small change in XrdOssSys::Features too (this is squashed in, it's inside one of the existing commits). That concerned Apple builds; where I didn't implement cloning (no FICLONE ioctl). However it can still honor the kxr_samefs option, so to allow that to work I set XRDOSS_HASFICL for the Oss independent of the availability of the ioctl. (The clone() methods will appropriately return not supported). [Mac OS does have clonefile, and variants, but I didn't think to try adding it. It works slightly differently to the ioctl, but could probably be made to work for whole file, but not range cloning].

abh3 commented on src/XrdOss/XrdOssConfig.cc:
```diff
@@ -2166,6 +2166,7 @@ void XrdOssSys::List_Path(const char *pfx, const char *pname,
          ss += (flags & XRDEXP_PURGE    ? " purge"   : " nopurge");
          ss += (flags & XRDEXP_RCREATE  ? " rcreate" : " norcreate");
          ss += (flags & XRDEXP_STAGE    ? " stage"   : " nostage");
+         ss += (flags & XRDEXP_NOFICL   ? " noficl"  : " ficl");
```
Could you write up any new configurable. Nothing fancy, just what you added. and either include it the pull request or send it to me or both.

abh3 commented on src/XrdOuc/XrdOucCloneSeg.hh:
```diff
@@ -0,0 +1,57 @@
+       XrdOfsFile* ofsFile; // Pointer to source file at Ofs level
+       XrdOssDF*   ossFile; // Pointer to source file at Oss level
+       int64_t     unxFD;   // The source file descriptor
+      } src;
```
Well you don't as it's all context driven. I came up with two alternatives, std::variant or the grand simplification that just requires an fd entry, no unions needed.

abh3 commented on src/XrdOss/XrdOssConfig.cc:
```diff
@@ -567,7 +567,7 @@ void XrdOssSys::ConfigCache(XrdSysError &Eroute, bool pass2)
 //
    if (DirFlags & XRDEXP_PFCACHE)
-      {DirFlags |=  XRDEXP_READONLY;
+      {DirFlags |=  XRDEXP_READONLY | XRDEXP_NOFICL;
```
I suppose that's the safest default.

amadio commented on src/XrdOssCsi/XrdOssCsi.hh:
```diff
@@ -76,32 +76,32 @@ friend class XrdOssCsiFileAio;
 
-virtual int     Close(long long *retsz=0) /* override */;
-virtual int     Open(const char *, int, mode_t, XrdOucEnv &) /* override */;
+virtual int     Close(long long *retsz=0) override;
```
@abh3 This does look unrelated to `FICLONE`, but a separate commit is ok, it doesn't have to be a full separate pull request. The guideline is [one logical change per commit](https://github.com/xrootd/xrootd/blob/master/docs/CONTRIBUTING.md#make-separate-commits-for-logically-separate-changes).

smithdh commented:
Hi. Pushed; rebase to latest master, fix few conficts during that. Thanks for latest review. I hope I've fixed all Andy's first round chage request/suggestions. (Going for 'grand simplification' route). Altough still missing is requested doc change concerning new config options (for space/export). Could you remind me where the doc source I should write or diff against is?

abh3 commented on src/XrdOss/XrdOssApi.cc:
```diff
@@ -254,6 +259,77 @@ int XrdOssSys::GenRemotePath(const char *oldp, char *newp)
+  do{
+   int sFd = cVec[j].srcFD;
+
+   for (i = j+1; i < n && cVec[i].srcFD == sFd; i++);
```
Given that the ioctl() can only do one element at a time, I don't think you need anything more that a simple loop that goes through the coloc vector and issues the ioctl() for each element. So, this can be dramatically simplified by using a single simple iterative loop.

abh3 commented on src/XrdOss/XrdOssApi.cc:
```diff
@@ -254,6 +259,77 @@ int XrdOssSys::GenRemotePath(const char *oldp, char *newp)
+       fr.src_length = cVec[k].srcLen;
+       fr.dest_offset = cVec[k].dstOffs;
+       int rc = ioctl(fd, FICLONERANGE, &fr);
+       if (rc < 0) return -errno;
```
If you are worried that you can't issue a reasonable message upon the error, use the new getErrMsg() API associated with the target file to relay context. Now that I said that, I see that was only fully implemented for XrdPss but with file cloning we need to implement it for XrdOss so that error text can be returned as simple errno's don't cut it any more. Further the commit that uses that API sits abandoned in the devel branch and really needs to move to master, sigh. This needs to be straightened out but it should stop the simplificaton.

smithdh commented:
I've reaslied that the current code for passing XrdSfsp->FName() via "oss.coloc" CGI will fail if the filename has an '&' in it, the name will be effectively truncated. I had a look around but I don't think we've got any established practice of encoding or escaping the value to avoid this in the Xrootd partions of our code, but I could take the XrdHttp approach (i.e. quote() from XrdHttpUtils.cc). But I'd like to check if there's a alternative suggesiton of how to handle this. Do you see any?

smithdh commented:
Hi Andy. I think I've addressed everything from the previous reviews; but sorry if I've missed something. Do you think you could re-review to see if there's more?

abh3 reviewed (approved):
Looks like all my concerns have been addressed.
